src/fixit/rules/deprecated_abc_import.py

```diff
@@ -0,0 +1,250 @@
+
+    def __init__(self) -> None:
+        super().__init__()
+        # If the module needs to updated
+        self.update_module = False
+
+    def visit_ImportFrom(self, node: cst.ImportFrom) -> None:
+        """
+        This catches the `from collections import <ABC>` cases
+        """
+        # Get imports in this statement
+        import_names = (
+            [name.name.value for name in node.names]
+            if type(node.names) is tuple
+            else []
+        )
+        # Filter the imports for ABC imports
+        import_names_in_abc = [name in ABCS for name in import_names]
+        if (
+            node.module
+            and node.module.value == "collections"
+            and any(import_names_in_abc)
+        ):
+            # Replacing the case where there are ABCs mixed with non-ABCs requires
+            # splitting a single import statement into two separate imports. This
+            # cannot be achieved in this method and is offloaded to leaving the module.
+            if not all(import_names_in_abc):
+                # We set this variable which triggers the `self.report` to be called
+                # in `leave_Module`. We report in the `leave_Module`
+                # so that we can add an additional `SimpleStatementLine` for the new
+                # import
+                self.update_module = True
+                self.imports_names = import_names
+            else:
+                self.report(
+                    node,
+                    replacement=node.with_changes(
+                        module=cst.Attribute(
+                            value=cst.Name(value="collections"),
+                            attr=cst.Name(value="abc"),
+                        )
+                    ),
+                )
+
+    def get_import_from(
+        self, node: Union[SimpleStatementLine, BaseCompoundStatement]
+    ) -> Optional[ImportFrom]:
+        """
+        Iterate over a Statement Sequence and return a Statement if it is a
+        `cst.ImportFrom` statement.
+        """
+        if isinstance(node, SimpleStatementLine):
+            for statement in node.body:
+                # Get the imports in this statement
+                import_names = (
+                    [
+                        name.name.value if isinstance(name, ImportAlias) else None
+                        for name in statement.names
+                    ]
+                    if isinstance(statement, ImportFrom)
+                    and isinstance(statement.names, tuple)
+                    else []
+                )
+                if (
+                    isinstance(statement, ImportFrom)
+                    and statement.module
+                    and statement.module.value == "collections"
+                    # Ensure the imports in this statement are the ones we are looking for
+                    and set(import_names).intersection(set(self.imports_names))
+                ):
+                    return statement
```

Comment:
It should be possible to express most of this with a libcst matcher and reduce the number of instance checks, etc. Something like:

```
import libcst.matchers as m

if m.matches(node, m.SimpleStatementLine(body=m.ZeroOrMore(m.ImportFrom(...)))):
    ...
```

Something like `m.findall(...)` may also be helpful.

https://libcst.readthedocs.io/en/latest/matchers_tutorial.html

https://libcst.readthedocs.io/en/latest/matchers.html